src/qt/walletmodel.cpp

```diff
@@ -291,12 +291,11 @@ bool WalletModel::updateAddressBookLabels(const CTxDestination& dest, const std:
 {
     LOCK(wallet->cs_wallet);
 
-    std::map<CTxDestination, AddressBook::CAddressBookData>::iterator mi = wallet->mapAddressBook.find(dest);
-
+    auto optAdd = pwalletMain->GetAddressBookEntry(dest);
     // Check if we have a new address or an updated label
-    if (mi == wallet->mapAddressBook.end()) {
+    if (!optAdd) {
         return wallet->SetAddressBook(dest, strName, strPurpose);
-    } else if (mi->second.name != strName) {
+    } else if (optAdd->name != strName) {
         return wallet->SetAddressBook(dest, strName, ""); // "" means don't change purpose
```

Comment:
This makes the LOCK at line 292 not needed now. Both `GetABEntry` and `SetAddressBook` lock cs_wallet when needed.